perf(parser): use 8 bits for each `Token` flag

overlookmotel opened the pull request:
Use 8 bits for each of the boolean flags in `Token`, so they can be read/written as `bool`s, avoiding bitwise operations.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11046?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11046** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11046?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11046?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11041** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11041?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#11042](https://github.com/oxc-project/oxc/pull/11042) <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11042?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#11040** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11040?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11046).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Small positive effect on parser benchmarks.

Boshen commented on crates/oxc_parser/src/lexer/token.rs:
```diff
@@ -109,42 +113,74 @@ impl Token {
+        // SAFETY: 8 bits starting at `ESCAPED_SHIFT` are only set in `Token::default` and
+        // `Token::set_escaped`. Both only set these bits to 0 or 1, so valid to read as a `bool`.
+        unsafe {
+            *ptr::from_ref(self).cast::<bool>().add(ESCAPED_SHIFT / 8).as_ref().unwrap_unchecked()
```
lol wtf am I reading?

Boshen commented:
### Merge activity

* **May 15, 5:29 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 15, 5:29 AM EDT**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
perf(parser): use 8 bits for each `Token` flag (#11046)

Use 8 bits for each of the boolean flags in `Token`, so they can be read/written as `bool`s, avoiding bitwise operations.
